### ISSUE TITLE
Inheriting custom base class for Controller crashes swagger

### ISSUE THREAD

**Issue by damircolak** [domaindrivendev/Swashbuckle.AspNetCore#730]
This:

public class UserController : MyClass.DbController

Crashes swagger.

This:

public class UserController : Controller

Works.

So, swagger cannot work if we inherit custom base clases?

![image](https://user-images.githubusercontent.com/17295346/39956648-39ae249e-55bb-11e8-8e5e-5568f0f473d1.png)

**Comment by domaindrivendev** [comment]
Navigate directly to the Swagger endpoint (/swagger/v1/s/wagger.json) and post the exception message

**Comment by aaxelm** [comment]
I noticed 2 issues regarding this.

1. When in a generic base controller there are 2 methods with the same name, an AmbiguousMatchException is thrown. If the second method is renamed or if the xml comments are not used in Swagger, everything works fine so a simple remporary fix for this in projects would be to have all methods with unique names in base generic controllers. Ex;

```
[Route("api/home")]
public class HomeController : BaseController<TestDto, int>
{
}

public abstract class BaseController<TDto, TKey> : Controller
    where TDto : class
{
    [HttpGet]
    public IActionResult Get()
    {
        return this.Ok();
    }

    [HttpGet("{id}")]
    public IActionResult Get(int id)
    {
        return this.Ok();
    }
}
```

2. The second scenario is also in a generic base controller, when one of the parameters of a method is also generic. In this case, a NullReferenceException is thrown. One temporary fix for this is to not use "IncludeXmlComments" in projects until this issue is fixed.
```
[Route("api/home")]
public class HomeController : BaseController<TestDto, int>
{
}

public abstract class BaseController<TDto, TKey> : Controller
    where TDto : class
{
    [HttpPatch]
    public virtual IActionResult Patch(TKey id, [FromBody] JsonPatchDocument<TDto> patchDoc)
    {
        return this.Ok();
    }
}
```
As for a fix in Swashbuckle, from what I was able to test, changing the `FullNameSansTypeArguments` method from the `Swashbuckle.AspNetCore.SwaggerGen.TypeExtensions` class does the trick:
```
internal static string FullNameSansTypeArguments(this Type type)
{
    if (string.IsNullOrEmpty(type.FullName))
    {
        return string.Empty;
    }

    var fullName = type.FullName;
    var chopIndex = fullName.IndexOf("[[");
    return (chopIndex == -1) ? fullName : fullName.Substring(0, chopIndex);
}
```

**Comment by domaindrivendev** [comment]
@damircolak - if you don't provide the exception details I can't help you. I'll wait another couple of days and close this issue if I don't here back.

@aaxelm - I'm not convinced that your findings re generics are necessarily relevant for this issue. Let's wait and see if @damircolak provides any further information before we decide how to track your discovery.

**Comment by domaindrivendev** [comment]
Closing as issue doesn't contain enough information to classify. @damircolak if you post the additional details I've requested above, I'll be happy to reopen.

**Comment by slubowsky** [comment]
@domaindrivendev Can confirm that having a generic base controller with multiple Get methods and then creating multiple controllers that  inherit from it causes `System.Reflection.AmbiguousMatchException: Ambiguous match found` error if you try to use IncludeXmlComments.

Easy workaround as mentioned above is to not have multiple methods in base with same name.  (e..g in example above rename second to GetById) - but not obvious at all when you run into this error what the cause is (no hint in the error message which controller is the problem...)
Full error message below:
```
System.Reflection.AmbiguousMatchException: Ambiguous match found.
   at System.RuntimeType.GetMethodImplCommon(String name, Int32 genericParameterCount, BindingFlags bindingAttr, Binder binder, CallingConventions callConv, Type[] types, ParameterModifier[] modifiers)
   at System.RuntimeType.GetMethodImpl(String name, BindingFlags bindingAttr, Binder binder, CallingConventions callConv, Type[] types, ParameterModifier[] modifiers)
   at System.Type.GetMethod(String name, BindingFlags bindingAttr)
   at System.Reflection.TypeExtensions.GetMethod(Type type, String name)
   at Swashbuckle.AspNetCore.SwaggerGen.XmlCommentsMemberNameHelper.GetMemberNameForMethod(MethodInfo method)
   at Swashbuckle.AspNetCore.SwaggerGen.XmlCommentsOperationFilter.Apply(Operation operation, OperationFilterContext context)
   at Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenerator.CreateOperation(ApiDescription apiDescription, ISchemaRegistry schemaRegistry)
   at Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenerator.CreatePathItem(IEnumerable`1 apiDescriptions, ISchemaRegistry schemaRegistry)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
   at Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenerator.CreatePathItems(IEnumerable`1 apiDescriptions, ISchemaRegistry schemaRegistry)
   at Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenerator.GetSwagger(String documentName, String host, String basePath, String[] schemes)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Builder.RouterMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.Invoke(HttpContext context)
```

**Comment by danieldegtyarev** [comment]
Please, reopen it. I got the same exception for my API

**Comment by abhaydhar** [comment]
I am also facing the same issue after migration from asp.net core 1.1 to 2.0 and swagger 3.0. Have raised on SO - https://stackoverflow.com/questions/51714557/no-operations-defined-in-spec-swashbuckle-aspnetcore

**Comment by abhaydhar** [comment]
Hi @domaindrivendev  - Do you need repro steps ? I hope you find the info i put on SO question as useful .

**Comment by domaindrivendev** [comment]
I've improved the behavior with 7baa8583fbc0bef4351f22a6a4cff62350f702f2 to handle this a little better. It's not perfect, but it's the best I can do right now.

I was unable to find a reliable method for retrieving the underlying generic type method from a constructed type method. I ended up settling on a best effort approach that looks for a unique match by name and parameter count. If there's multiple matches (hopefully an edge case), then it will fail gracefully. Here's the relevant method for reference:

```csharp
private MethodInfo GetGenericTypeMethodOrNullFor(MethodInfo constructedTypeMethod)
{
    var constructedType = constructedTypeMethod.DeclaringType;
    var genericTypeDefinition = constructedType.GetGenericTypeDefinition();

    // Retrieve list of candidate methods that match name and parameter count
    var candidateMethods = genericTypeDefinition.GetMethods()
        .Where(m =>
        {
            return (m.Name == constructedTypeMethod.Name)
                && (m.GetParameters().Length == constructedTypeMethod.GetParameters().Length);
        });
         
    // If inconclusive, just return null
    return (candidateMethods.Count() == 1) ? candidateMethods.First() : null;
}
```

This will be available with the upcoming 4.0.0 release (date still TBD). In the meantime, you can pull down the preview package from myget.org:

https://www.myget.org/feed/domaindrivendev/package/nuget/Swashbuckle.AspNetCore
